[Dang Quang Thang] iP

quangthangisme opened the pull request:
# 🦆 MightyDuck 🦆

Do you often forget your tasks? MightyDuck to the rescue! He is
- text-based
- easy to learn
- **SUPER FAST** to use

Just follow the steps
1. Download MightyDuck from [here](https://github.com/quangthangisme/ip/)
2. Tell him your tasks
3. And let him manage them!

He comes with
- [x] A user-friendly ~~CLI~~ GUI
- [x] A task managing system
- [ ] Reminders _(Coming soon!)_

> MightyDuck to the rescue!

Here's a code snippet from `Launcher`
```java
public static void main(String[] args) {
    Application.launch(Main.class, args);
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

quangthangisme committed:
Level-0: Rename, Greet, Exit

quangthangisme committed:
Level-1: Echo

quangthangisme committed:
Level-2: Add, List

quangthangisme committed:
Level-3: Mark as Done

quangthangisme committed:
Level-4: ToDo, Event, Deadline

quangthangisme committed:
A-TextUiTesting: Automated Text UI Testing

quangthangisme committed:
Level-5: Handle Errors

quangthangisme committed:
Level-6: Delete

quangthangisme committed:
Level-7: Save

quangthangisme committed:
Merge branch 'branch-Level-7'

quangthangisme committed:
Level-8: Dates and Times

quangthangisme committed:
Merge branch 'branch-Level-8'

quangthangisme committed:
A-MoreOOP: Use More OOP

quangthangisme committed:
Merge branch 'branch-A-MoreOOP'

quangthangisme committed:
A-Packages: Organize into Packages

quangthangisme committed:
Merge branch 'branch-A-Packages'

quangthangisme committed:
Merge remote-tracking branch 'origin/add-gradle-support'

quangthangisme committed:
A-Gradle: Use Gradle

quangthangisme committed:
A-JUnit: Add JUnit Tests

quangthangisme committed:
Merge branch 'branch-A-JUnit'

quangthangisme committed:
Add JavaDoc

quangthangisme committed:
Comply with the given coding standard

quangthangisme committed:
Add find functionality

quangthangisme committed:
Merge branch 'branch-A-JavaDoc'

quangthangisme committed:
Merge branch 'branch-A-CodingStandard'

quangthangisme committed:
Merge branch 'branch-Level-9'

quangthangisme committed:
Add JavaDoc and fix style for Level-9

quangthangisme committed:
Remove unnecessary usages of keyword this

quangthangisme committed:
Update Ui to handle printing blank lines

quangthangisme committed:
Build a JavaFx GUI

quangthangisme committed:
Merge branch 'Level-10'

quangthangisme committed:
Refactor command parsing and creation logic into builder classes

GrassHeadd commented on src/main/java/mightyduck/command/Parser.java:
```diff
@@ -0,0 +1,147 @@
+        case ByeCommand.COMMAND_WORD:
+            return new ByeCommand(taskManager);
+        case ListCommand.COMMAND_WORD:
+            return new ListCommand(taskManager);
+        case MarkCommand.COMMAND_WORD: {
+            int index = parseIndex(argumentsStr);
+            return new MarkCommand(taskManager, index);
+        }
+        case UnmarkCommand.COMMAND_WORD: {
+            int index = parseIndex(argumentsStr);
+            return new UnmarkCommand(taskManager, index);
+        }
+        case DeleteCommand.COMMAND_WORD: {
+            int index = parseIndex(argumentsStr);
+            return new DeleteCommand(taskManager, index);
+        }
+        case ToDoCommand.COMMAND_WORD: {
+            String[] todoArguments = parseArguments(argumentsStr, ToDoCommand.COMMAND_FORMAT,
+                    ToDoCommand.KEYWORDS);
+            return new ToDoCommand(taskManager, todoArguments);
+        }
+        case DeadlineCommand.COMMAND_WORD: {
+            String[] dlArguments = parseArguments(argumentsStr, DeadlineCommand.COMMAND_FORMAT,
+                    DeadlineCommand.KEYWORDS);
+            return new DeadlineCommand(taskManager, dlArguments);
+        }
+        case EventCommand.COMMAND_WORD: {
+            String[] eventArguments = parseArguments(argumentsStr, EventCommand.COMMAND_FORMAT,
+                    EventCommand.KEYWORDS);
+            return new EventCommand(taskManager, eventArguments);
+        }
+        case FindCommand.COMMAND_WORD: {
+            String[] findArguments = parseArguments(argumentsStr, FindCommand.COMMAND_FORMAT,
+                    FindCommand.KEYWORDS);
+            return new FindCommand(taskManager, findArguments);
+        }
+        default:
+            throw new InvalidCommandException(Messages.INVALID_COMMAND);
+        }
+    }
```
seems like a very long method, could consider splitting them up having a main parser, then call different methods base on the input

quangthangisme replied:
Thanks! I realized that and refactored the Parser class. Constructing the commands from the user's input will now be delegated to Builder classes.

GrassHeadd commented on src/main/java/mightyduck/command/commands/ListCommand.java:
```diff
@@ -0,0 +1,51 @@
+        List<Task> tasks = taskManager.getTasks();
+        return new CommandResult(
+                CommandResultType.SUCCESS,
+                tasks.isEmpty() ? Messages.EMPTY_LIST : Messages.LIST,
+                IntStream.range(0, tasks.size())
+                        .mapToObj(i -> new Pair<>(i, tasks.get(i)))
+                        .toList()
```
like the splitting of overly long lines

GrassHeadd commented on src/main/java/mightyduck/command/Parser.java:
```diff
@@ -0,0 +1,147 @@
+        for (int i = 0; i < keywords.size(); i++) {
+            int nextIndex = argumentsStr.indexOf(keywords.get(i));
+            if (nextIndex == -1) {
+                throw new InvalidCommandException(
+                        String.format(Messages.WRONG_COMMAND_FORMAT, format));
```
something i realise on my own machine is that if you throw the exception, it will terminate the program, which in general not desired for a chatbot, maybe could return a sout to just print out a message that you have the wrong format, but im not sure how this works out on ur program, just a suggestion!

quangthangisme replied:
In my implementation, if an `InvalidCommandException` is thrown, then `MightyDuck` will catch that and return a `CommandResult` of type `CommandResultType.ERROR`. The controller will fetch the feedback message from the `CommandResult` and display the error message.

GrassHeadd commented on src/main/java/mightyduck/command/Parser.java:
```diff
@@ -0,0 +1,147 @@
+        case ByeCommand.COMMAND_WORD:
+            return new ByeCommand(taskManager);
+        case ListCommand.COMMAND_WORD:
+            return new ListCommand(taskManager);
+        case MarkCommand.COMMAND_WORD: {
+            int index = parseIndex(argumentsStr);
+            return new MarkCommand(taskManager, index);
+        }
+        case UnmarkCommand.COMMAND_WORD: {
+            int index = parseIndex(argumentsStr);
+            return new UnmarkCommand(taskManager, index);
+        }
+        case DeleteCommand.COMMAND_WORD: {
+            int index = parseIndex(argumentsStr);
+            return new DeleteCommand(taskManager, index);
+        }
+        case ToDoCommand.COMMAND_WORD: {
+            String[] todoArguments = parseArguments(argumentsStr, ToDoCommand.COMMAND_FORMAT,
+                    ToDoCommand.KEYWORDS);
+            return new ToDoCommand(taskManager, todoArguments);
+        }
+        case DeadlineCommand.COMMAND_WORD: {
+            String[] dlArguments = parseArguments(argumentsStr, DeadlineCommand.COMMAND_FORMAT,
+                    DeadlineCommand.KEYWORDS);
+            return new DeadlineCommand(taskManager, dlArguments);
+        }
+        case EventCommand.COMMAND_WORD: {
+            String[] eventArguments = parseArguments(argumentsStr, EventCommand.COMMAND_FORMAT,
+                    EventCommand.KEYWORDS);
+            return new EventCommand(taskManager, eventArguments);
+        }
+        case FindCommand.COMMAND_WORD: {
+            String[] findArguments = parseArguments(argumentsStr, FindCommand.COMMAND_FORMAT,
+                    FindCommand.KEYWORDS);
+            return new FindCommand(taskManager, findArguments);
+        }
+        default:
+            throw new InvalidCommandException(Messages.INVALID_COMMAND);
+        }
+    }
```
seems like a very long method, could consider splitting them up having a main parser, then call different methods base on the input

GrassHeadd commented on src/main/java/mightyduck/command/CommandResultType.java:
```diff
@@ -0,0 +1,28 @@
+    SUCCESS,
+
+    /**
+     * Indicates an error result due to incorrect user input or other invalid command issues.
+     */
+    ERROR,
```
maybe can change the name to be less abstract rather than error like maybe INVALID_INPUT_ERROR for better debugging

quangthangisme replied:
I will consider that!

GrassHeadd commented on src/main/java/mightyduck/messages/Messages.java:
really like the messages:)

quangthangisme replied:
Thanks! I'm glad someone likes my cheesy puns =)))

GrassHeadd reviewed (commented):
really really well done, honestly not much i could point out. 
whole program is really orderly. 
Good job!

JadeCheah commented on src/main/java/mightyduck/command/commands/FindCommand.java:
```diff
@@ -0,0 +1,74 @@
+            throw new InvalidValueException(String.format(
+                    Messages.WRONG_COMMAND_FORMAT, COMMAND_FORMAT));
+        }
+        List<Pair<Integer, Task>> res = taskManager.searchKeyword(arguments[0]);
```
I'm not exactly sure what res, does it meant as result?

quangthangisme replied:
Yes, it means result. The variable may be better named as result.

JadeCheah commented on src/main/java/mightyduck/controller/MainController.java:
```diff
@@ -0,0 +1,115 @@
+package mightyduck.controller;
+
+import javafx.application.Platform;
+import javafx.fxml.FXML;
+import javafx.scene.control.Alert;
+import javafx.scene.control.Alert.AlertType;
+import javafx.scene.control.Button;
+import javafx.scene.control.ScrollPane;
+import javafx.scene.control.TextField;
+import javafx.scene.image.Image;
+import javafx.scene.layout.AnchorPane;
+import javafx.scene.layout.VBox;
+import mightyduck.MightyDuck;
+import mightyduck.command.CommandResult;
+import mightyduck.command.CommandResultType;
+import mightyduck.data.task.Task;
+import mightyduck.exception.InvalidStoragePathException;
+import mightyduck.exception.StorageLoadException;
+import mightyduck.utils.Pair;
```
Maybe can try adding a blank line between the javafx and mightyduck import statements to improve readability?
```suggestion
import javafx.application.Platform;
import javafx.fxml.FXML;
import javafx.scene.control.Alert;
import javafx.scene.control.Alert.AlertType;
import javafx.scene.control.Button;
import javafx.scene.control.ScrollPane;
import javafx.scene.control.TextField;
import javafx.scene.image.Image;
import javafx.scene.layout.AnchorPane;
import javafx.scene.layout.VBox;

import mightyduck.MightyDuck;
import mightyduck.command.CommandResult;
import mightyduck.command.CommandResultType;
import mightyduck.data.task.Task;
import mightyduck.exception.InvalidStoragePathException;
import mightyduck.exception.StorageLoadException;
import mightyduck.utils.Pair;
```

quangthangisme replied:
I followed the tutorial on how to configure the code style on IntelliJ.

![image](https://github.com/user-attachments/assets/9884c262-f59a-4926-9540-66e676a1dbe8)

Based on the image, I believe there is no harm in omitting the blank line. But thanks regardless!

JadeCheah reviewed (commented):
Everything aligns perfectly with the coding standards. Variable, function, and class names are clear, easily understood, and readable. Switch-case indentations are aligned with the recommended ones. Overall, exceptionally done.

quangthangisme committed:
Fix some typos in JavaDoc comments

quangthangisme committed:
Add assertions to validate expected behavior

There are several implicit assumptions in the code, such as a
TaskManager passed into a Command being non-null. By adding assertions
at key points, we make these assumptions explicit and ensure the
expected behavior is clearly defined. Using assertions is preferable to
exceptions in this case, as they allow us to detect issues early in
development without introducing the overhead of exception handling.

quangthangisme committed:
Improve code quality

There are some code quality violations that hinder readability and
maintainability.

To address these issues, we will:
* Refactor long methods into smaller, more manageable functions to
improve readability.
* Replace magic numbers and strings with constants for better
maintainability.
* Rename variables to be more descriptive and better reflect their
purpose, enhancing clarity.

quangthangisme committed:
Merge pull request #1 from quangthangisme/branch-A-Assertions

Add assertions to validate expected behavior

quangthangisme committed:
Merge branch 'master' into branch-A-CodeQuality

quangthangisme committed:
Merge pull request #2 from quangthangisme/branch-A-CodeQuality

Improve code quality

quangthangisme committed:
Remove automatic exit after "bye" command

The application automatically exits after the "bye" command.

After testing, we find this behavior too sudden and awkward.

To improve the user experience, we disable the user input field and the
submit button instead of exiting the application after the "bye"
command.

quangthangisme committed:
Add tagging and untagging features

quangthangisme committed:
Enable commands to handle multiple indices

quangthangisme committed:
Merge branch 'branch-BCD-Extension'

quangthangisme committed:
Set up Continuous Integration

quangthangisme committed:
Fix some header comments

quangthangisme committed:
Improve the GUI

quangthangisme committed:
Add declaration of AI usage

quangthangisme committed:
Add an icon and title to the application

quangthangisme committed:
Add user guide

quangthangisme committed:
Update Config class to include storage path
